Major refactoring of msgpack rcp layer, support for subscribe/unsubscribe/broadcast (issue 60)

stze opened the pull request:
This PR introduces a refactored msgpack rpc layer and adds support for API calls subscribe/unsubscribe and broadcast.

mkind committed:
incrementing version number

Commit:
Add memdup functions to util module

* memdup without nul-terminate
* memdup with nul-terminate

Commit:
Replace message serialize/deserialize and pack/unpack

The msgpack helper functions implemented in msgpack/helper.c do the
serializing and deserializing of msgpack-rcp messages from or to the
object structure. This adds also the support for msgpack dictionary
type.

The api helper functions implemented in api/helpers.c help when
constructing/freeing a object-message structure.

Commit:
Major api function simplification using the new helper modules

Commit:
Include missing NULL checks

Commit:
Remove tests from old module and adapt tests to new helper modules

This commit removes tests (pack/unpack/message) which are now deprecated
because of the new helper module. I also tried to simplify the
run/register/result tests and i fixed every memleak produced from these
tests.

Commit:
Refactor rpc modules and add support for subscribe/unsubscribe/broadcast

This commit cleans up the rpc stack and introduces an event
subscribe/unsubscribe and broadcast mechanism.

Commit:
Update CMakeLists.txt

stze commented on src/rpc/msgpack/message.c:
```diff
@@ -1,469 +0,0 @@
-/**
```
this commit also deletes deprecated tests (see commit  fb6b70aa8f22bfb4e9b758cfaa7d152af28050b2).

Commit:
Add functional tests for subscribe and unsubscribe

Commit:
Revert "Add functional tests for subscribe and unsubscribe"

This reverts commit 7d4f4af7fc466b7fd09f05b2bf9eea11f688c7ce.

Commit:
Add functional tests for subscribe/unsubscribe

Commit:
Add functional test for broadcast API call

Commit:
Make request event handler more generic

Commit:
Add more functional tests of broadcast API call

Commit:
Add functional tests for new helper modules.

Commit:
Add additional tests to msgpack helper functional test

Commit:
Pre-increment msgid before sending packet

Commit:
Fix overwriting connection in hashmap

bontric commented:
There are currently three problems I observed:
1. A function  is currently required to have a description (This is not specified!). This can be fixed by automatically adding an empty docstring to any function in the client library or adding a check at the core:
   function.c:167

``` c
// Add check: if desc == NULL 
  if (db_function_add_meta(pluginkey, name, desc) == -1)
    return (-1);
```
1. If a plugin tries to register with an invalid function description the core simply doesn't store the function but registers the plugin anyways and sends a "unable to register function" error response. (register.c:41-55). The plugin should be removed from the database as a consequence.
2.  Run calls don't work properly. I couldn't find the problem yet but a call is not forwarded to the target plugin and no response is sent. However, there is no error at the core. 

```
//core debugging output:
generated callid 127709182498857
sending request: method = run,  callinfo id = 2
```

mkind commented:
Is this pr work in progess? If so, I would add the **[WIP]** tag to the title.

Commit:
Include msgpack/libuv static cmake build scripts from neovim

Commit:
Switch to multiqueue neovim implementation and refactor uv loop handling

This commit has lots of changes, but actually only the event queue
implementation changed to the multiqueue implementation from neovim
(which is more generic). Second change is a refactoring of the event
loop implementation, which was implemented in unusable way and now is
easy usable.

Commit:
Gardening with include what you use

Commit:
Fix bug which produced wrong responses

Commit:
Add missing header files

Commit:
Try to make travis happy and downgrade msgpack from 2.0.0 to 1.4.2

Commit:
Remove unused unit tests

Commit:
Revert "Try to make travis happy and downgrade msgpack from 2.0.0 to 1.4.2"

This reverts commit 123465bc2860a7e832bd98a22b0de5b186e74487.

Commit:
Try to make travis happy and downgrade msgpack version from 2.0.0 to
1.0.0

Commit:
Update travis config to new output dir structure

Commit:
Fix some gcc wshadow warnings

Commit:
Add python 3 to travis yml

Commit:
Add travis before install script to update python pip

Commit:
Fix git permission for travis before install script

Commit:
Rearange travis before install dependencies

Commit:
Fix git permissions for travis install script

Commit:
Try to fix travis cmocka install failure

Commit:
Try to fix gcov cmake flag usage

mkind committed:
The check is not needed since the pointer is valid by definition. This
holds because it's only used in non-public manner.

mkind committed:
adding sb assertions to api-calls

mkind committed:
moving declaration at top of function

mkind committed:
increasing version number to 0.1.7

mkind committed:
Merge branch 'master' into 2016-10-stze-issue-60

mkind committed:
Merge branch '2016-10-stze-issue-60' of github.com:splone/splonebox-core into 2016-10-stze-issue-60
